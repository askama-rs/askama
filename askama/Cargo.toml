[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
