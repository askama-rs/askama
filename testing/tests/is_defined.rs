use askama::Template;

// This test ensures that `include` are correctly working inside filter blocks and that external
// variables are used correctly.
#[test]
fn is_defined_in_expr() {
    #[derive(Template)]
    #[template(
        source = r#"<script>
const x = {{ x is defined }};
const y = {{ y is not defined }};
const z = {{ y is defined }};
const w = {{ x is not defined }};
const v = {{ y }};
</script>"#,
        ext = "html"
    )]
    struct IsDefined {
        y: u32,
    }

    let s = IsDefined { y: 0 };
    assert_eq!(
        s.render().unwrap(),
        r"<script>
const x = false;
const y = false;
const z = true;
const w = true;
const v = 0;
</script>"
    );
}

// This test ensures that if the variable is not defined, it will not generate following code.
#[test]
fn is_defined_chaining() {
    #[derive(Template)]
    #[template(
        source = r#"{% if x is defined && x == 12 %}bli{% else %}bla{% endif %}"#,
        ext = "html"
    )]
    struct IsDefinedChaining;

    assert_eq!(IsDefinedChaining.render().unwrap(), r"bla");
}
