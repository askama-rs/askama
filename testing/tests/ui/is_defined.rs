use askama::Template;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if x.y is defined %}{% endif %}"#,
)]
struct A;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if true is defined %}{% endif %}"#,
)]
struct B;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if true is %}{% endif %}"#,
)]
struct C;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if x is %}{% endif %}"#,
)]
struct D;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if x is blue %}{% endif %}"#,
)]
struct E;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if x is blue.red %}{% endif %}"#,
)]
struct F;

fn main() {
}
