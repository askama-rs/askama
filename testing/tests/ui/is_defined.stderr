error: `is defined` operator can only be used on variables, not on their fields
 --> <source attribute>:1:10
       "is defined %}{% endif %}"
 --> tests/ui/is_defined.rs:6:14
  |
6 |     source = r#"{% if x.y is defined %}{% endif %}"#,
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `is defined` operator can only be used on variables
 --> <source attribute>:1:11
       "is defined %}{% endif %}"
  --> tests/ui/is_defined.rs:13:14
   |
13 |     source = r#"{% if true is defined %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: expected `defined` or `not defined` after `is`
 --> <source attribute>:1:11
       "is %}{% endif %}"
  --> tests/ui/is_defined.rs:20:14
   |
20 |     source = r#"{% if true is %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: expected `defined` or `not defined` after `is`
 --> <source attribute>:1:8
       "is %}{% endif %}"
  --> tests/ui/is_defined.rs:27:14
   |
27 |     source = r#"{% if x is %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: expected `defined` or `not defined` after `is`
 --> <source attribute>:1:8
       "is blue %}{% endif %}"
  --> tests/ui/is_defined.rs:34:14
   |
34 |     source = r#"{% if x is blue %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: expected `defined` or `not defined` after `is`
 --> <source attribute>:1:8
       "is blue.red %}{% endif %}"
  --> tests/ui/is_defined.rs:41:14
   |
41 |     source = r#"{% if x is blue.red %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
